runtime/src/main/java/org/corfudb/runtime/clients/NettyClientProtobufRouter.java

```diff
@@ -0,0 +1,360 @@
+
+    /** The {@link CorfuRuntime.CorfuRuntimeParameters} used to configure the
+     *  router.
+     */
+    private final RuntimeParameters parameters;
+
+    /**
+     * The outstanding requests on this router.
+     */
+    public final Map<Long, CompletableFuture> outstandingRequests;
+
+    /** A {@link CompletableFuture} which is completed when a connection,
+     *  including a successful handshake completes and messages can be sent
+     *  to the remote node.
+     */
+    @Getter
+    volatile CompletableFuture<Void> connectionFuture;
+
+    /**
+     * Timer map for measuring request
+     */
+    private final Map<MessageType, String> timerNameCache;
+
+    /**
+     * Add a new client to the router
+     *
+     * @param client The client to add to the router
+     * @return This IClientProtobufRouter
+     */
+    public IClientProtobufRouter addClient(IClient client) {
+        // TODO: Rewrite IClient and corresponding methods,
+
+        return this;
+    }
+
+    /**
+     * Send a request message and get a completable future to be fulfilled by the reply.
+     *
+     * @param request The request message to send.
+     * @param ctx Context of the channel handler.
+     * @param <T> The type of completable to return.
+     * @return A completable future which will be fulfilled by the reply,
+     * or a timeout in the case there is no response.
+     */
+    @Override
+    public  <T> CompletableFuture<T> sendRequestAndGetCompletable(Request request, ChannelHandlerContext ctx) {
+
+        // Check the connection future. If connected, continue with sending the message.
+        // If timed out, return a exceptionally completed with the timeout.
+        try {
+            connectionFuture
+                    .get(parameters.getConnectionTimeout().toMillis(), TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+            throw new UnrecoverableCorfuInterruptedError(e);
+        } catch (TimeoutException te) {
+            CompletableFuture<T> f = new CompletableFuture<>();
+            f.completeExceptionally(te);
+            return f;
+        } catch (ExecutionException ee) {
+            CompletableFuture<T> f = new CompletableFuture<>();
+            f.completeExceptionally(ee.getCause());
+            return f;
+        }
+
+        // Set up the timer and context to measure request
+        final Timer roundTripMsgTimer = CorfuRuntime.getDefaultMetrics()
+                .timer(timerNameCache.get(message.getMsgType()));
+
+        final Timer.Context roundTripMsgContext = MetricsUtils
+                .getConditionalContext(roundTripMsgTimer);
+
+        // Get the next request ID
+        final long thisRequest = requestID.getAndIncrement();
+
+        // Set the base fields for this message.
+        Request.Builder reqBuilder = request.toBuilder();
+        reqBuilder.getHeaderBuilder().setClientId(parameters.getClientId()).setRequestId(thisRequest);
+        request = reqBuilder.build();
```

Comment:
At this point, the Request is already built by the client calling this method. Calling toBuilder() involves making a copy of the whole object (I believe including payload), which seems wasteful just to update some fields in the header. I'm wondering if this can be avoided... In a perfect world, we would only build the request once. Perhaps if the Request can't be built entirely by the client, it could be better to pass a builder() and delay calling build()?

Reply:
Yeah that's also what I was thinking about, pass a builder as the parameter could avoid building it twice. I will make the change in next commit.